js/vendor/sd-streams.esm.js

```diff
@@ -0,0 +1,2484 @@
+            shutDown(() => readableStreamCancel(source, error), { actualError: error });
+        }
+        else {
+            shutDown(undefined, { actualError: error });
+        }
+    });
+    onStreamClosed(source, reader[closedPromise_].promise, () => {
+        if (!preventClose) {
+            shutDown(() => writableStreamDefaultWriterCloseWithErrorPropagation(writer));
+        }
+        else {
+            shutDown();
+        }
+    });
+    if (writableStreamCloseQueuedOrInFlight(dest) || dest[state_] === "closed") {
+        const destClosed = new TypeError();
+        if (!preventCancel) {
+            shutDown(() => readableStreamCancel(source, destClosed), { actualError: destClosed });
+        }
+        else {
+            shutDown(undefined, { actualError: destClosed });
+        }
+    }
+    function awaitLatestWrite() {
+        const curLatestWrite = latestWrite;
+        return latestWrite.then(() => curLatestWrite === latestWrite ? undefined : awaitLatestWrite());
+    }
+    function flushRemainder() {
+        if (dest[state_] === "writable" && (!writableStreamCloseQueuedOrInFlight(dest))) {
+            return awaitLatestWrite();
+        }
+        else {
+            return undefined;
+        }
+    }
+    function shutDown(action, error) {
+        if (shuttingDown) {
+            return;
+        }
+        shuttingDown = true;
+        if (action === undefined) {
+            action = () => Promise.resolve();
+        }
+        function finishShutDown() {
+            action().then(_ => finalize(error), newError => finalize({ actualError: newError }));
+        }
+        const flushWait = flushRemainder();
+        if (flushWait) {
+            flushWait.then(finishShutDown);
+        }
+        else {
+            finishShutDown();
+        }
+    }
+    function finalize(error) {
+        writableStreamDefaultWriterRelease(writer);
+        readableStreamReaderGenericRelease(reader);
+        if (signal && abortAlgorithm) {
+            signal.removeEventListener("abort", abortAlgorithm);
+        }
+        if (error) {
+            promise.reject(error.actualError);
+        }
+        else {
+            promise.resolve(undefined);
+        }
+    }
+    function next() {
+        if (shuttingDown) {
+            return;
+        }
+        writer[readyPromise_].promise.then(() => {
+            readableStreamDefaultReaderRead(reader).then(({ value, done }) => {
+                if (done) {
+                    return;
+                }
+                latestWrite = writableStreamDefaultWriterWrite(writer, value).catch(() => { });
+                next();
+            }, _error => {
+                latestWrite = Promise.resolve();
+            });
+        });
+    }
+    if (!shuttingDown) {
+        next();
+    }
+    return promise.promise;
+}
+
+class ReadableStreamDefaultController {
+    constructor() {
+        throw new TypeError();
+    }
+    get desiredSize() {
+        ret
```

Comment:
Wow this is a lot of code... Is all of this necessary to expose Request?

Reply:
Yeah it's intense.

Most of it is necessary to support `req.clone()` and `resp.clone()`, which both use `stream.tee()`.

This is mostly why I want you all looking at it early, it's basically going from some interfaces to a complete whatwg streams implementation. You could arguably avoid the transform stream code but I think dropping any of that makes `fetch` less useful.

Reply:
I don't completely understand how `/third_party` works, but would it make more sense to include this there?

Reply:
I went ahead and moved that dependency to third_party.

Reply:
@mrkurt Is it on npm? If so we could just add it to package.json (which would put it in third_party/node_modules) but I'd be fine if we kept it in //js - the JS code that is included in the runtime is more special that other dependencies, because we're really shipping it. So I kinda prefer actually forking the code having control over it.

Reply:
@ry it is! It's actually really nice TypeScript with an MIT license, so we could pull the source in directly: https://github.com/stardazed/sd-streams/tree/master/packages/streams/src